### PR TITLE
callbacks for dask optimization loop

### PR DESCRIPTION
**Pull request by sfalkner**
- add pre- and post-evaluation callbacks to dask optimization loop
- update mypy due to bug with numpy code

### CODE REVIEW

File: blackboxopt/optimization_loops/dask_distributed.py [hunk 1]
```diff
@@ -164,6 +174,11 @@ def run_optimization_loop(
                 break
 
         new_evaluations = dask_scheduler.check_for_results(timeout_s=20)
+
+        if post_evaluation_callback:
+            for e in new_evaluations:
+                post_evaluation_callback(e)
```
Comment (dynobo) [inline blackboxopt/optimization_loops/dask_distributed.py:180]: What do you think about
```suggestion
        map(post_evaluation_callback, new_evaluations)
```
or
```suggestion
        _ = map(post_evaluation_callback, new_evaluations)
```
?

**Review by dynobo — APPROVED** [review]
Thanks, @sfalkner , seems like a very useful feature addition :-)

File: blackboxopt/optimization_loops/dask_distributed.py (hunk 1 above)
Comment (dynobo) [reply blackboxopt/optimization_loops/dask_distributed.py:180]: I messed up this suggestion. I'll create a new comment.

Comment (dynobo) [inline blackboxopt/optimization_loops/dask_distributed.py:180]: What do you think about this instead?
```suggestion
            list(map(post_evaluation_callback, new_evaluations))
```

**Review by LGro — APPROVED** [review]
For the sequential loop we have pre and post evaluation callback tests and I think this would also be a good idea here to ensure that we don't accidentally break this API and also to demonstrate the expected behaviour that the order of the evaluations is not necessarily consistently ordered.

**Pull request merged**